plugins/Live/Model.php

```diff
@@ -112,6 +115,84 @@ public function queryLogVisits($idSite, $period, $date, $segment, $offset, $limi
         return $foundVisits;
     }
 
+    private function executeLogVisitsQuery($sql, $bind, $segment, $dateStart, $dateEnd, $minTimestamp, $limit)
+    {
+        $readerDb = Db::getReader();
+        try {
+            $visits = $readerDb->fetchAll($sql, $bind);
+        } catch (Exception $e) {
+            // we also need to check for the 'maximum statement execution time exceeded' text as the query might be
+            // aborted at different stages and we can't really know all the possible codes at which it may be aborted etc
+            $isMaxExecutionTimeError = $readerDb->isErrNo($e, DbMigration::ERROR_CODE_MAX_EXECUTION_TIME_EXCEEDED_QUERY_INTERRUPTED)
+                || $readerDb->isErrNo($e, DbMigration::ERROR_CODE_MAX_EXECUTION_TIME_EXCEEDED_SORT_ABORTED)
+                || strpos($e->getMessage(), 'maximum statement execution time exceeded') !== false;
+
+            if ($isMaxExecutionTimeError) {
+                $message = '';
+
+                if ($this->isLookingAtMoreThanOneDay($dateStart, $dateEnd, $minTimestamp)) {
+                    $message .= ' ' . Piwik::translate('Live_QueryMaxExecutionTimeExceededReasonDateRange');
+                }
+
+                if (!empty($segment)) {
+                    $message .= ' ' . Piwik::translate('Live_QueryMaxExecutionTimeExceededReasonSegment');
+                }
+
+                $limitThatCannotBeSelectedInUiButOnlyApi = 550;
+                if ($limit > $limitThatCannotBeSelectedInUiButOnlyApi) {
+                    $message .= ' ' . Piwik::translate('Live_QueryMaxExecutionTimeExceededLimit');
+                }
+
+                if (empty($message)) {
+                    $message .= ' ' . Piwik::translate('Live_QueryMaxExecutionTimeExceededReasonUnknown');
+                }
+
+                $message = Piwik::translate('Live_QueryMaxExecutionTimeExceeded') . ' ' . $message;
+
+                $params = array(
+                    'sql' => $sql, 'bind' => $bind, 'segment' => $segment, 'limit' => $limit
+                );
+
+                /**
+                 * @ignore
+                 * @internal
+                 */
+                Piwik::postEvent('Live.queryMaxExecutionTimeExceeded', array($params));
+                throw new MaxExecutionTimeExceededException($message);
+            }
+            throw $e;
+        }
+        return $visits;
+    }
```

Comment:
This code seems important. Would it be possible to unit/integration test this code w/ a mock Db object returned by `Db::getReader()`?